### PR TITLE
[Snapshots] Fix device specs

### PR DESCRIPTION
**Pull request by rbro112**
Fixes device specs to be true-to-size to reference devices. Also adjusts to properly scale custom specs. Added a repro case of a troublesome client snapshot that wasn't rendering in landscaped.

Verified with a sxs run against Google's compose preview snapshotting library, Emerge matches 1:1 on all device specs now. Android Studio images (copied directly from the previews) appear smaller, and dimensions do not match the reference device specs, so I believe matching the expected device dimensions like we do now is preferred over 1:1 matching of AS.

Also checked all reference devices against the outputted dimensions from Emerge snapshots after this change and all match 1:1.

Screenshot 2024-10-22 at 9 15 45 AM

### CODE REVIEW

**Comment by rbro112** [comment]
* **#285** <a href="https://app.graphite.dev/github/pr/EmergeTools/emerge-android/285?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rbro112 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/ComposeSnapshotter.kt [hunk 1]
```diff
@@ -170,87 +178,88 @@ private fun snapshot(
...
+  val width = deviceSpec.widthPixels
+  val height = deviceSpec.heightPixels
+
+  if (width > 0 && height > 0) {
+    activity.window.setLayout(width, height)
+  }
+}
+
+private fun dpToPx(dp: Int, scalingFactor: Float): Int {
+  return (dp * scalingFactor).toInt()
```
Comment (trevor-e) [inline snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/ComposeSnapshotter.kt:260]: I'm guessing this just needs to be explicitly rounded down.

File: snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/DeviceUtils.kt [hunk 2]
```diff
@@ -256,43 +210,24 @@ fun configToDeviceSpec(config: ComposePreviewSnapshotConfig): DeviceSpec? {
...
+        widthDp = height
       }
-      DimensionSpec(
-        widthDp = widthDp,
-        heightDp = heightDp,
-        densityPpi = densityPpi,
-        scalingFactor = densityPpi?.let(DeviceSpec::getClosestDensityScalingFactor)
+
+      DeviceSpec(
+        heightDp = heightDp ?: -1,
```
Comment (trevor-e) [inline snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/DeviceUtils.kt:224]: Is it possible to get rid of these -1 representations everywhere and instead use nil?

**Review by trevor-e — APPROVED** [review]

File: snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/ComposeSnapshotter.kt (hunk 1 above)
Comment (rbro112) [reply snapshots/snapshots/src/main/kotlin/com/emergetools/snapshots/compose/ComposeSnapshotter.kt:260]: I updated to use `toInt()` in the pixel conversion later. Ended up with the proper dimensions going with that approach

**Pull request merged**